### ISSUE TITLE
Layers

### ISSUE THREAD

**Issue by petrasl1976** [excalidraw/excalidraw#7725]
Hi,

Could we add layers to Excalidraw for organizing objects? Being able to show or hide these layers would make it perfect for step-by-step presentations, offering more control than typical slides by grouping elements for each topic or chapter.

Best regards,
PetrasL

**Comment by CWempe** [comment]
I really would love layers, too.

But have a look at https://github.com/excalidraw/excalidraw/issues/2170.

**Comment by teghnet** [comment]
Yes. Layers. :pray: 
If we could stack/embed independent excalidraw diagrams, that would also work, but they would need to be live docs.

**Comment by teghnet** [comment]
I wholeheartedly agree with the key points.

**Comment by OldStarchy** [comment]
I agree with all the points too, especially 3., in my case I'm looking in particular something that can toggle layer groups on an off for a data flow presentation. A decision tree with individual paths annotated on different layers would allow for a crude but very effective and easy to use slideshow-esc tool for running through different scenarios during meetings while allowing live editing.

The lack of this feature is the reason why (in this situation) I need to find a different tool.

**Comment by clombion** [comment]
The presentation workflow is also what brought me here. I often use excalidraw to represent technical systems, and adding or removing details from a base schema is a powerful way to tell a story about a system. Currently I have to duplicate the each schema multiple times if I want to do this kind of storytelling. 

Even beyond presentation, being able to make complex drawings without having to lock/unlock objects multiple times in the process would be very valuable.

**Comment by anvinh** [comment]
Working in Excalidraw would be more simple and faster if layers are possible:
- I need layers to manage my system design drawing. Or any large drawing.
- Because the feature does not exist, I have to "imitate" it with grouping, which is not easy if you have up to 10 elements spreading up bottom left right in your landscape -> it slows me down  

Please add layer feature. Basically it's around 4 key functionalities + related UI: 
- create/delete a layer
- manage the layers: move up/down
- merge all/n-selected layers

**Comment by johannes-z** [comment]
I'm trying to implement layers on top of `<Excalidraw />` and while there are a few hacks, nothing seems really viable right now. I came up with these possibilities:

1. Lock all elements outside of the current layer - Downside: they can still be selected. You can prevent unlock with the onChange event, but you can't prevent selection.
2. Use opacity for elements in a layer that's hidden. Downside: they can still be selected...

Another possibility would be manipulating the canvas directly, but the API doesn't really allow to do this:
1. When switching layers, copy the current canvas state and save it for later.
2. use `api.updateScene` to remove all elements not in the active layer.
3. draw the copied canvas layer on top of or below the current canvas elements.

The problem is, that there is no way to render elements _behind_ the current scene, only on top. I also tried creating a second canvas that overlays with excalidraw, but this had issues with transparency as well.

---

### A possible solution

When inspecting the source code, what's really missing is the following element metadata that's handled respectively:
`visibility`:
* `active` or undefined for normal elements, same behavior as now
* `static` skips hit scan - can't be interactived with, e.g. clicked, moved, etc. only changable by api.
* `invisible`: skips hit scan and rendering.

I think this would need to change:

```typescript
type _ExcalidrawElementBase = Readonly<{
  // ...
  visibility?: 'static' | 'active' | 'invisible' 
}>

// App.tsx line 5197
private getElementAtPosition ...
elements.filter((el) => !el.visibility || el.visibility === 'active')

// renderElement.ts line 729
const renderElement ...
if (element.visibility === 'invisible') return

// selection.ts line 67
export const getElementsWithinSelection
let elementsInSelection = elements.filter(e => !e.visibility || e.visibility === 'active')...
```
+some other methods (for cursor state, etc.)

The layer information, i.e. which element is in which layer can be done using the element's customData, e.g. `customData: { layerId: '<guid>' }` and the visibility state can be set in the `onChange` callback.

I'm pretty sure that's all we need to implement layers in userland, but I haven't tried this yet.
Additionally, if the `ExcalidrawImperativeAPI` exposed a method to select item(s) by id, you could do even more features, e.g. when clicking a layer, select all elements in the layer.

These changes would be minimal, non-breaking and effectively allow consuming apps to implement layers, which would not interfere with excalidraw's general design at all.

**Comment by Rkdroid** [comment]
Hey everyone! 👋

I've been following this issue for a while and finally decided to take a shot at implementing it.

**PR: #10447**

### What's Implemented

**Layer CRUD Operations**
- Create new layers via "+" button
- Rename layers (double-click to edit)
- Delete layers (elements move to default layer)
- Reorder layers to control z-index stacking

**Visibility Control**
- Toggle layer visibility with eye icon
- Hidden layers' elements are not rendered or selectable

**Element Management**
- New elements are created on the active layer
- "Move to Layer" context menu option
- Element count displayed per layer

**Multi-Select & Merge**
- Select multiple layers with Cmd/Ctrl + Click
- Merge selected layers into one
- "Merge All" to combine all layers

**Backward Compatibility**
- Existing files without layer data work seamlessly
- Elements default to the first layer automatically

---

🔗 **Live demo:** https://excalidraw-3bjnanfey-excalidraw.vercel.app

![Image](https://github.com/user-attachments/assets/3097f305-0d44-4596-8f64-8b58ba2c5d35)

---

I'd love feedback from the maintainers and community on:
- Does this align with the project's direction?
- Any UX improvements needed?
- Edge cases I might have missed?

Happy to iterate based on feedback. Thanks!

**Comment by petrasl1976** [comment]
Hey! Thats amazing! I really like it. Thank you so much!

**Comment by anvinh** [comment]
It's fire @Rkdroid

**Comment by BI-Squirrel** [comment]
@Rkdroid  It's great.
 
Really like the count of elements per Layer at a first glance, maybe it's weird vor huge diagrams.

I missed the possibility to copy/cut-paste selected elements from one layer to another.

**Comment by t3dodson** [comment]
I hope this feature is reconsidered to be added to excalidraw. I think this functionality is so fundamental to how users expect drawing apps to work that this feature could cause the community to fracture and adopt a fork with this functionality.

**Comment by Zikoel** [comment]
Just my opinion, I think this can be a killer feature for excalidraw from some reasons:
- the feature is not invasive, who don't want layers, can simply ignore it
- other tools starts to offer this feature (miro added this in 2025 I think)
- cover a lot of use-cases.

Here an example of use-case that my team need every day:

We have a lot of infrastructure schema and when we need to explain that to our collaborator is obsolutely awesom to have layers for tecnical comments, user data connections, control plane data connections, logs flow, authentication flows etc. Every concept can be assigned to a layer an simplify a lot every view of the same schema. The current alternative is create a duplication of the entire schema multiple times.

A feature for layers that can be interesting to add is "lock layer" a layer locked will lock all its elements, this permit to work freely on a layer leaving unouthed another layer.
